src/main/java/net/fabricmc/loom/transformers/CompiledJarRemappingTransformer.java

```diff
@@ -0,0 +1,144 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.transformers;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.zeroturnaround.zip.ZipUtil;
+import org.zeroturnaround.zip.commons.FileUtils;
+import org.gradle.api.artifacts.transform.CacheableTransform;
+import org.gradle.api.artifacts.transform.InputArtifact;
+import org.gradle.api.artifacts.transform.InputArtifactDependencies;
+import org.gradle.api.artifacts.transform.TransformAction;
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.Project;
+import org.gradle.api.tasks.CompileClasspath;
+import org.apache.tools.ant.util.StreamUtils;
+
+import net.fabricmc.loom.LoomGradleExtension;
+import net.fabricmc.loom.providers.MappingsProvider;
+import net.fabricmc.loom.providers.MinecraftMappedProvider;
+import net.fabricmc.loom.transformers.parameters.ProjectReferencingParameters;
+import net.fabricmc.loom.util.FabricModUtils;
+import net.fabricmc.loom.util.ModProcessor;
+import net.fabricmc.loom.util.TinyRemapperMappingsHelper;
+import net.fabricmc.tinyremapper.OutputConsumerPath;
+import net.fabricmc.tinyremapper.TinyRemapper;
+
+@CacheableTransform
+public abstract class CompiledJarRemappingTransformer implements TransformAction<ProjectReferencingParameters> {
+	@CompileClasspath
+	@InputArtifact
+	public abstract Provider<FileSystemLocation> getInput();
+
+	@CompileClasspath
+	@InputArtifactDependencies
+	abstract FileCollection getDependencies();
+
+	@Override
+	public void transform(final TransformOutputs outputs) {
+		final FileSystemLocation resolvedInput = getInput().getOrNull();
+		if (resolvedInput == null) {
+			lifecycle("Can not transform a jar who has no input.");
+			return;
+		}
+
+		final File inputFile = resolvedInput.getAsFile();
+		if (!inputFile.exists()) {
+			lifecycle("Can not transform jar who does not exists on disk.");
+			return;
+		}
+
+		if (!FabricModUtils.isFabricMod(
+						getProject(),
+						getProject().getLogger(),
+						inputFile,
+						inputFile.getName()
+		)) {
+			final File outputFile = outputs.file("not-remapped-no-mod" + File.separator + inputFile.getName());
+			try {
+				FileUtils.copyFile(inputFile, outputFile);
+			} catch (IOException e) {
+				lifecycle("Failed to copy not remappable file to output.", e);
+			}
+            return;
+		}
+
+		//Rudimentary check for sources. Which need to be remapped separately.
+        //Potentially we need to check zip file contents.
+		if (inputFile.getName().contains("-sources"))
```

Comment:
Yep totally agree.
Some lambda logic and a builder would probably go a long way.
But I would love to write this up a bit better so that it works in nearly all cases perfectly before I go mayham on cleaning loom up and making it more stable.